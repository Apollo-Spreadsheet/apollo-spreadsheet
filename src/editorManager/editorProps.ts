import { StopEditingParams } from './useEditorManager'
import { Column } from '../columnGrid'
import { ApiRef } from '../api'

export interface EditorProps {
  value: string
  stopEditing: (params?: StopEditingParams) => void
  anchorRef: Element
  additionalProps: Column['editorProps']
  maxLength: number
  validatorHook?: Column['validatorHook']
  apiRef: ApiRef
}
