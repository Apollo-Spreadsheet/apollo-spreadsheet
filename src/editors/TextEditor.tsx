import React, { CSSProperties, useRef, useState } from "react";
import {
  Popover,
  TextField,
  TextareaAutosize,
  Fade,
  Grow,
} from "@material-ui/core";
import { NavigationKey } from ".//navigation-key.enum";
import { isCaretAtEndPosition } from "./utils";

interface Props {
  value: string;
  onCommit: (value: string, navigationKey?: NavigationKey) => void;
  onCommitCancel: (navigationKey?: NavigationKey) => void;
  anchorRef: Element;
  cellWidth: React.ReactText;
  cellHeight: React.ReactText;
  maxLength: number;
}
const textAreaStyle: CSSProperties = {
  width: "100%",
  height: "100%",
  resize: "none",
  overflow: "auto",
};


export function TextEditor({
  value,
  onCommit,
  onCommitCancel,
  anchorRef,
  cellWidth,
  cellHeight,
  maxLength,
}: Props) {
  const changedValue = useRef<string>(value);
  const onKeyDown = (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
    /**
     * @todo Needs a review and moved into base editor for better support
     */
    if (e.key === "Enter" || NavigationKey[e.key]) {
      const cursorStart = e.target["selectionStart"];
      const cursorEnd = e.target["selectionEnd"];
      // console.log({
      //   cursorStart,
      //   cursorEnd,
      // });
      //Only navigate if we are in the end
      if (
        (e.key === NavigationKey.ArrowRight || NavigationKey.ArrowDown) &&
        !isCaretAtEndPosition(cursorEnd, e.target["value"].length)
      ) {
        return;
      }
      //Only navigate if we are in the left
      if (
        (e.key === NavigationKey.ArrowLeft ||
          e.key === NavigationKey.ArrowUp) &&
        cursorStart > 0
      ) {
        return console.warn("YAY");
      }

      e.preventDefault();
      dispatchUpdate(NavigationKey[e.key]);
    }
  };

  const dispatchUpdate = (navigationKey?: NavigationKey) => {
    if (changedValue.current !== value) {
      return onCommit(changedValue.current, navigationKey);
    }

    return onCommitCancel(navigationKey);
  };

  return (
    <Popover
      anchorEl={anchorRef}
      open
      elevation={1}
      TransitionProps={{ timeout: 0 }}
      onClose={(event, reason) => {
        if (reason === "backdropClick") {
          return dispatchUpdate();
        }
        onCommitCancel();
      }}
      anchorOrigin={{
        vertical: "center",
        horizontal: "center",
      }}
      transformOrigin={{
        vertical: "center",
        horizontal: "center",
      }}
      PaperProps={{
        style: {
          overflow: "hidden",
        },
      }}
    >
      <div style={{ width: cellWidth, height: cellHeight }}>
        <TextareaAutosize
          defaultValue={changedValue.current}
          onChange={(e) => (changedValue.current = e.target["value"])}
          autoFocus={true}
          onKeyDown={onKeyDown}
          aria-label="cell editor"
          rowsMin={1}
          maxLength={maxLength ?? 500}
          style={textAreaStyle}
        />
      </div>
    </Popover>
  );
}

export default TextEditor;
