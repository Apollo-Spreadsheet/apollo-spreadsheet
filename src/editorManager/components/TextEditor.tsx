import React, {
	CSSProperties,
	forwardRef,
	useCallback,
	useEffect,
	useImperativeHandle,
	useMemo,
	useRef,
	useState,
} from 'react'
import { Popover, TextField, TextareaAutosize, Fade, Grow } from '@material-ui/core'
import { NavigationKey } from '../enums/navigation-key.enum'
import { isCaretAtEndPosition } from '../utils/isCaretAtEndPosition'
import { EditorProps } from '../editorProps'
import { makeStyles } from '@material-ui/core/styles'

const useStyles = makeStyles(() => ({
	input: {
		width: '100%',
		height: '100%',
		resize: 'none',
		overflow: 'auto',
		border: 0,
		outline: 0,
		'&:focus': {
			border: 0,
			outline: 0,
		},
	},
}))
export const TextEditor = forwardRef(
	({ value, stopEditing, anchorRef, maxLength, validatorHook }: EditorProps, componentRef) => {
		const classes = useStyles()
		const [editingValue, setEditingValue] = useState(String(value))
		useImperativeHandle(
			componentRef,
			() => ({
				getValue: () => {
					return editingValue
				},
			}),
			[editingValue],
		)

		const isValidValue = useMemo(() => {
			if (!validatorHook) {
				return true
			}
			return validatorHook(editingValue)
		}, [editingValue])

		const onKeyDown = (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
			if (NavigationKey[e.key]) {
				const cursorStart = e.target['selectionStart']
				const cursorEnd = e.target['selectionEnd']
				console.log({
					key: e.key,
					cursorStart,
					cursorEnd,
				})
				if (cursorStart === 0) {
					console.error('Cursor at start')
					e.preventDefault()
					stopEditing({ save: true, keyPress: NavigationKey[e.key] })
				}

				if (isCaretAtEndPosition(cursorEnd, e.target['value'].length)) {
					console.error('Cursor at end')
					e.preventDefault()
					stopEditing({ save: true, keyPress: NavigationKey[e.key] })
				}
				// console.log({
				// 	cursorStart,
				// 	cursorEnd,
				// 	textLength: e.target['value'].length,
				// 	isCaretAtEndPosition: isCaretAtEndPosition(cursorEnd, e.target['value'].length),
				// })
				//
				// const cartInEnd = isCaretAtEndPosition(cursorEnd, e.target['value'].length)
				// //Only navigate if we are in the end
				// if (!cartInEnd) {
				// 	return console.error("At end")
				// }
				//
				// //Check if caret is not yet trying to go negative or zero
				// if (cursorStart > 0 && cursorEnd > 0) {
				// 	return console.log("oh yea")
				// }
				//
				// e.preventDefault()
				// stopEditing({ save: true, keyPress: NavigationKey[e.key] })
			}
		}

		const onTextAreaResizeMount = useCallback((ref: HTMLTextAreaElement | null) => {
			if (!ref) {
				return
			}
			ref.selectionStart = editingValue.length
			ref.selectionEnd = editingValue.length
		}, [])

		console.log({ editingValue })
		const anchorStyle = anchorRef['style'] as CSSProperties
		return (
			<Popover
				id={'editor-portal'}
				anchorEl={anchorRef}
				open
				elevation={0}
				TransitionProps={{ timeout: 0 }}
				onClose={(event, reason) => {
					//Only allow to cancel if its invalid
					if (!isValidValue) {
						return stopEditing({ save: false })
					}

					if (reason === 'backdropClick') {
						return stopEditing({ save: true })
					}

					stopEditing({ save: false })
				}}
				marginThreshold={10}
				disableRestoreFocus
				PaperProps={{
					style: {
						overflow: 'hidden',
						zIndex: 10,
						border: isValidValue ? anchorStyle.border : '1px solid red',
						borderRadius: 0,
					},
				}}
			>
				<div
					id="editor-container"
					style={{
						width: anchorStyle.width,
						minHeight: anchorStyle.height,
					}}
				>
					<TextareaAutosize
						id={'apollo-textarea'}
						value={editingValue}
						ref={onTextAreaResizeMount}
						onChange={e => setEditingValue(e.target['value'])}
						autoFocus
						onKeyDown={onKeyDown}
						aria-label="text apollo editor"
						rowsMin={1}
						maxLength={maxLength}
						className={classes.input}
					/>
				</div>
			</Popover>
		)
	},
)

export default TextEditor
