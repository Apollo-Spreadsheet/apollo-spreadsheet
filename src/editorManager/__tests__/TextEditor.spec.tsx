import { mount, shallow } from 'enzyme'
import TextEditor from '../components/TextEditor'
import { EditorProps } from '../editorProps'
import React from 'react'

describe('<TextEditor />', () => {
	const onCommitMock = jest.fn(() => {
	})
	const onCommitCancelMock = jest.fn(() => {
	})
	// const mockAnchor = document.createElement('div')
	const props: EditorProps = {
		value: '',
		onCommit: onCommitMock,
		onCommitCancel: onCommitCancelMock,
		anchorRef: document.body,
		cellWidth: 100,
		cellHeight: 50,
		maxLength: 500,
	}
	const wrapper = mount(<TextEditor {...props} />)
	it('should match snapshot', () => {
		expect(wrapper).toMatchSnapshot()
	})

	it('should close on ESC', () => {
		console.log(wrapper.find('#apollo-textarea'))
		wrapper.find('#apollo-textarea').simulate('keydown', { key: 'Escape' })
		console.log()
	})
})