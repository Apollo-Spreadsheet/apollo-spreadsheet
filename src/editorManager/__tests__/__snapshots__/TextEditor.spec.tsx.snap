// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextEditor /> should match snapshot 1`] = `
<TextEditor
  anchorRef={
    <body
      style="padding-right: 0px; overflow: hidden;"
    >
      <div
        class="MuiPopover-root"
        role="presentation"
        style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
      >
        <div
          aria-hidden="true"
          style="z-index: -1; position: fixed; right: 0px; bottom: 0px; top: 0px; left: 0px; background-color: transparent;"
        />
        <div
          data-test="sentinelStart"
          tabindex="0"
        />
        <div
          class="MuiPaper-root MuiPopover-paper MuiPaper-elevation1 MuiPaper-rounded"
          style="opacity: 1; transform: scale(1, 1); overflow: hidden; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
          tabindex="-1"
        >
          <div
            style="width: 100px; height: 50px;"
          >
            <textarea
              aria-label="cell editor"
              id="apollo-textarea"
              maxlength="500"
              rows="1"
              style="height: 100%; overflow: auto; width: 100%; resize: none;"
            />
            <textarea
              aria-hidden="true"
              readonly=""
              style="visibility: hidden; position: absolute; overflow: auto; height: 100%; top: 0px; left: 0px; transform: translateZ(0); width: 100%; resize: none;"
              tabindex="-1"
            />
          </div>
        </div>
        <div
          data-test="sentinelEnd"
          tabindex="0"
        />
      </div>
    </body>
  }
  cellHeight={50}
  cellWidth={100}
  maxLength={500}
  onCommit={[MockFunction]}
  onCommitCancel={[MockFunction]}
  value=""
>
  <WithStyles(ForwardRef(Popover))
    PaperProps={
      Object {
        "style": Object {
          "overflow": "hidden",
        },
      }
    }
    TransitionProps={
      Object {
        "timeout": 0,
      }
    }
    anchorEl={
      <body
        style="padding-right: 0px; overflow: hidden;"
      >
        <div
          class="MuiPopover-root"
          role="presentation"
          style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
        >
          <div
            aria-hidden="true"
            style="z-index: -1; position: fixed; right: 0px; bottom: 0px; top: 0px; left: 0px; background-color: transparent;"
          />
          <div
            data-test="sentinelStart"
            tabindex="0"
          />
          <div
            class="MuiPaper-root MuiPopover-paper MuiPaper-elevation1 MuiPaper-rounded"
            style="opacity: 1; transform: scale(1, 1); overflow: hidden; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
            tabindex="-1"
          >
            <div
              style="width: 100px; height: 50px;"
            >
              <textarea
                aria-label="cell editor"
                id="apollo-textarea"
                maxlength="500"
                rows="1"
                style="height: 100%; overflow: auto; width: 100%; resize: none;"
              />
              <textarea
                aria-hidden="true"
                readonly=""
                style="visibility: hidden; position: absolute; overflow: auto; height: 100%; top: 0px; left: 0px; transform: translateZ(0); width: 100%; resize: none;"
                tabindex="-1"
              />
            </div>
          </div>
          <div
            data-test="sentinelEnd"
            tabindex="0"
          />
        </div>
      </body>
    }
    anchorOrigin={
      Object {
        "horizontal": "center",
        "vertical": "center",
      }
    }
    disableRestoreFocus={true}
    elevation={1}
    onClose={[Function]}
    open={true}
    transformOrigin={
      Object {
        "horizontal": "center",
        "vertical": "center",
      }
    }
  >
    <ForwardRef(Popover)
      PaperProps={
        Object {
          "style": Object {
            "overflow": "hidden",
          },
        }
      }
      TransitionProps={
        Object {
          "timeout": 0,
        }
      }
      anchorEl={
        <body
          style="padding-right: 0px; overflow: hidden;"
        >
          <div
            class="MuiPopover-root"
            role="presentation"
            style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
          >
            <div
              aria-hidden="true"
              style="z-index: -1; position: fixed; right: 0px; bottom: 0px; top: 0px; left: 0px; background-color: transparent;"
            />
            <div
              data-test="sentinelStart"
              tabindex="0"
            />
            <div
              class="MuiPaper-root MuiPopover-paper MuiPaper-elevation1 MuiPaper-rounded"
              style="opacity: 1; transform: scale(1, 1); overflow: hidden; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
              tabindex="-1"
            >
              <div
                style="width: 100px; height: 50px;"
              >
                <textarea
                  aria-label="cell editor"
                  id="apollo-textarea"
                  maxlength="500"
                  rows="1"
                  style="height: 100%; overflow: auto; width: 100%; resize: none;"
                />
                <textarea
                  aria-hidden="true"
                  readonly=""
                  style="visibility: hidden; position: absolute; overflow: auto; height: 100%; top: 0px; left: 0px; transform: translateZ(0); width: 100%; resize: none;"
                  tabindex="-1"
                />
              </div>
            </div>
            <div
              data-test="sentinelEnd"
              tabindex="0"
            />
          </div>
        </body>
      }
      anchorOrigin={
        Object {
          "horizontal": "center",
          "vertical": "center",
        }
      }
      classes={
        Object {
          "paper": "MuiPopover-paper",
          "root": "MuiPopover-root",
        }
      }
      disableRestoreFocus={true}
      elevation={1}
      onClose={[Function]}
      open={true}
      transformOrigin={
        Object {
          "horizontal": "center",
          "vertical": "center",
        }
      }
    >
      <ForwardRef(Modal)
        BackdropProps={
          Object {
            "invisible": true,
          }
        }
        className="MuiPopover-root"
        container={
          <body
            style="padding-right: 0px; overflow: hidden;"
          >
            <div
              class="MuiPopover-root"
              role="presentation"
              style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
            >
              <div
                aria-hidden="true"
                style="z-index: -1; position: fixed; right: 0px; bottom: 0px; top: 0px; left: 0px; background-color: transparent;"
              />
              <div
                data-test="sentinelStart"
                tabindex="0"
              />
              <div
                class="MuiPaper-root MuiPopover-paper MuiPaper-elevation1 MuiPaper-rounded"
                style="opacity: 1; transform: scale(1, 1); overflow: hidden; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
                tabindex="-1"
              >
                <div
                  style="width: 100px; height: 50px;"
                >
                  <textarea
                    aria-label="cell editor"
                    id="apollo-textarea"
                    maxlength="500"
                    rows="1"
                    style="height: 100%; overflow: auto; width: 100%; resize: none;"
                  />
                  <textarea
                    aria-hidden="true"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: auto; height: 100%; top: 0px; left: 0px; transform: translateZ(0); width: 100%; resize: none;"
                    tabindex="-1"
                  />
                </div>
              </div>
              <div
                data-test="sentinelEnd"
                tabindex="0"
              />
            </div>
          </body>
        }
        disableRestoreFocus={true}
        onClose={[Function]}
        open={true}
      >
        <ForwardRef(Portal)
          container={
            <body
              style="padding-right: 0px; overflow: hidden;"
            >
              <div
                class="MuiPopover-root"
                role="presentation"
                style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
              >
                <div
                  aria-hidden="true"
                  style="z-index: -1; position: fixed; right: 0px; bottom: 0px; top: 0px; left: 0px; background-color: transparent;"
                />
                <div
                  data-test="sentinelStart"
                  tabindex="0"
                />
                <div
                  class="MuiPaper-root MuiPopover-paper MuiPaper-elevation1 MuiPaper-rounded"
                  style="opacity: 1; transform: scale(1, 1); overflow: hidden; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
                  tabindex="-1"
                >
                  <div
                    style="width: 100px; height: 50px;"
                  >
                    <textarea
                      aria-label="cell editor"
                      id="apollo-textarea"
                      maxlength="500"
                      rows="1"
                      style="height: 100%; overflow: auto; width: 100%; resize: none;"
                    />
                    <textarea
                      aria-hidden="true"
                      readonly=""
                      style="visibility: hidden; position: absolute; overflow: auto; height: 100%; top: 0px; left: 0px; transform: translateZ(0); width: 100%; resize: none;"
                      tabindex="-1"
                    />
                  </div>
                </div>
                <div
                  data-test="sentinelEnd"
                  tabindex="0"
                />
              </div>
            </body>
          }
          disablePortal={false}
        >
          <Portal
            containerInfo={
              <body
                style="padding-right: 0px; overflow: hidden;"
              >
                <div
                  class="MuiPopover-root"
                  role="presentation"
                  style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
                >
                  <div
                    aria-hidden="true"
                    style="z-index: -1; position: fixed; right: 0px; bottom: 0px; top: 0px; left: 0px; background-color: transparent;"
                  />
                  <div
                    data-test="sentinelStart"
                    tabindex="0"
                  />
                  <div
                    class="MuiPaper-root MuiPopover-paper MuiPaper-elevation1 MuiPaper-rounded"
                    style="opacity: 1; transform: scale(1, 1); overflow: hidden; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
                    tabindex="-1"
                  >
                    <div
                      style="width: 100px; height: 50px;"
                    >
                      <textarea
                        aria-label="cell editor"
                        id="apollo-textarea"
                        maxlength="500"
                        rows="1"
                        style="height: 100%; overflow: auto; width: 100%; resize: none;"
                      />
                      <textarea
                        aria-hidden="true"
                        readonly=""
                        style="visibility: hidden; position: absolute; overflow: auto; height: 100%; top: 0px; left: 0px; transform: translateZ(0); width: 100%; resize: none;"
                        tabindex="-1"
                      />
                    </div>
                  </div>
                  <div
                    data-test="sentinelEnd"
                    tabindex="0"
                  />
                </div>
              </body>
            }
          >
            <div
              className="MuiPopover-root"
              onKeyDown={[Function]}
              role="presentation"
              style={
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "fixed",
                  "right": 0,
                  "top": 0,
                  "zIndex": 1300,
                }
              }
            >
              <ForwardRef(SimpleBackdrop)
                invisible={true}
                onClick={[Function]}
                open={true}
              >
                <div
                  aria-hidden={true}
                  onClick={[Function]}
                  style={
                    Object {
                      "WebkitTapHighlightColor": "transparent",
                      "backgroundColor": "transparent",
                      "bottom": 0,
                      "left": 0,
                      "position": "fixed",
                      "right": 0,
                      "top": 0,
                      "zIndex": -1,
                    }
                  }
                />
              </ForwardRef(SimpleBackdrop)>
              <Unstable_TrapFocus
                disableAutoFocus={false}
                disableEnforceFocus={false}
                disableRestoreFocus={true}
                getDoc={[Function]}
                isEnabled={[Function]}
                open={true}
              >
                <div
                  data-test="sentinelStart"
                  tabIndex={0}
                />
                <ForwardRef(Grow)
                  appear={true}
                  in={true}
                  onEnter={[Function]}
                  onEntering={[Function]}
                  onExited={[Function]}
                  tabIndex="-1"
                  timeout={0}
                >
                  <Transition
                    addEndListener={[Function]}
                    appear={true}
                    enter={true}
                    exit={true}
                    in={true}
                    mountOnEnter={false}
                    onEnter={[Function]}
                    onEntered={[Function]}
                    onEntering={[Function]}
                    onExit={[Function]}
                    onExited={[Function]}
                    onExiting={[Function]}
                    tabIndex="-1"
                    timeout={0}
                    unmountOnExit={false}
                  >
                    <WithStyles(ForwardRef(Paper))
                      className="MuiPopover-paper"
                      elevation={1}
                      style={
                        Object {
                          "opacity": 1,
                          "overflow": "hidden",
                          "transform": "scale(1, 1)",
                          "visibility": undefined,
                        }
                      }
                      tabIndex="-1"
                    >
                      <ForwardRef(Paper)
                        className="MuiPopover-paper"
                        classes={
                          Object {
                            "elevation0": "MuiPaper-elevation0",
                            "elevation1": "MuiPaper-elevation1",
                            "elevation10": "MuiPaper-elevation10",
                            "elevation11": "MuiPaper-elevation11",
                            "elevation12": "MuiPaper-elevation12",
                            "elevation13": "MuiPaper-elevation13",
                            "elevation14": "MuiPaper-elevation14",
                            "elevation15": "MuiPaper-elevation15",
                            "elevation16": "MuiPaper-elevation16",
                            "elevation17": "MuiPaper-elevation17",
                            "elevation18": "MuiPaper-elevation18",
                            "elevation19": "MuiPaper-elevation19",
                            "elevation2": "MuiPaper-elevation2",
                            "elevation20": "MuiPaper-elevation20",
                            "elevation21": "MuiPaper-elevation21",
                            "elevation22": "MuiPaper-elevation22",
                            "elevation23": "MuiPaper-elevation23",
                            "elevation24": "MuiPaper-elevation24",
                            "elevation3": "MuiPaper-elevation3",
                            "elevation4": "MuiPaper-elevation4",
                            "elevation5": "MuiPaper-elevation5",
                            "elevation6": "MuiPaper-elevation6",
                            "elevation7": "MuiPaper-elevation7",
                            "elevation8": "MuiPaper-elevation8",
                            "elevation9": "MuiPaper-elevation9",
                            "outlined": "MuiPaper-outlined",
                            "root": "MuiPaper-root",
                            "rounded": "MuiPaper-rounded",
                          }
                        }
                        elevation={1}
                        style={
                          Object {
                            "opacity": 1,
                            "overflow": "hidden",
                            "transform": "scale(1, 1)",
                            "visibility": undefined,
                          }
                        }
                        tabIndex="-1"
                      >
                        <div
                          className="MuiPaper-root MuiPopover-paper MuiPaper-elevation1 MuiPaper-rounded"
                          style={
                            Object {
                              "opacity": 1,
                              "overflow": "hidden",
                              "transform": "scale(1, 1)",
                              "visibility": undefined,
                            }
                          }
                          tabIndex="-1"
                        >
                          <div
                            style={
                              Object {
                                "height": 50,
                                "width": 100,
                              }
                            }
                          >
                            <ForwardRef(TextareaAutosize)
                              aria-label="cell editor"
                              autoFocus={true}
                              defaultValue=""
                              id="apollo-textarea"
                              maxLength={500}
                              onChange={[Function]}
                              onKeyDown={[Function]}
                              rowsMin={1}
                              style={
                                Object {
                                  "height": "100%",
                                  "overflow": "auto",
                                  "resize": "none",
                                  "width": "100%",
                                }
                              }
                            >
                              <textarea
                                aria-label="cell editor"
                                autoFocus={true}
                                defaultValue=""
                                id="apollo-textarea"
                                maxLength={500}
                                onChange={[Function]}
                                onKeyDown={[Function]}
                                rows={1}
                                style={
                                  Object {
                                    "height": "100%",
                                    "overflow": "auto",
                                    "resize": "none",
                                    "width": "100%",
                                  }
                                }
                              />
                              <textarea
                                aria-hidden={true}
                                readOnly={true}
                                style={
                                  Object {
                                    "height": "100%",
                                    "left": 0,
                                    "overflow": "auto",
                                    "position": "absolute",
                                    "resize": "none",
                                    "top": 0,
                                    "transform": "translateZ(0)",
                                    "visibility": "hidden",
                                    "width": "100%",
                                  }
                                }
                                tabIndex={-1}
                              />
                            </ForwardRef(TextareaAutosize)>
                          </div>
                        </div>
                      </ForwardRef(Paper)>
                    </WithStyles(ForwardRef(Paper))>
                  </Transition>
                </ForwardRef(Grow)>
                <div
                  data-test="sentinelEnd"
                  tabIndex={0}
                />
              </Unstable_TrapFocus>
            </div>
          </Portal>
        </ForwardRef(Portal)>
      </ForwardRef(Modal)>
    </ForwardRef(Popover)>
  </WithStyles(ForwardRef(Popover))>
</TextEditor>
`;
